Framework/Indexing/src/IndexInfo.cpp

```diff
@@ -13,99 +14,72 @@ namespace Indexing {
 }
 
 /// Construct with given spectrum number and vector of detector IDs for each
 /// index.
-IndexInfo::IndexInfo(std::vector<SpectrumNumber> &&spectrumNumbers,
-                     std::vector<std::vector<DetectorID>> &&detectorIDs) {
-  if (spectrumNumbers.size() != detectorIDs.size())
-    throw std::runtime_error("IndexInfo: Size mismatch between spectrum number "
-                             "and detector ID vectors");
-  m_spectrumNumbers.access() = std::move(spectrumNumbers);
-  m_detectorIDs.access() = std::move(detectorIDs);
+IndexInfo::IndexInfo(std::vector<SpectrumNumber> spectrumNumbers,
+                     const StorageMode storageMode,
+                     const Communicator &communicator)
+    : m_storageMode(storageMode), m_communicator(communicator) {
+  makeSpectrumNumberTranslator(std::move(spectrumNumbers));
 }
 
 /// The *local* size, i.e., the number of spectra in this partition.
 size_t IndexInfo::size() const {
-  if (!m_spectrumNumbers)
+  if (!m_spectrumNumberTranslator)
     return 0;
-  return m_spectrumNumbers->size();
+  return m_spectrumNumberTranslator->localSize();
 }
 
-/// Returns the spectrum number for given index.
-SpectrumNumber IndexInfo::spectrumNumber(const size_t index) const {
-  return (*m_spectrumNumbers)[index];
+/// The *global* size, i.e., the total number of spectra across all partitions.
+size_t IndexInfo::globalSize() const {
+  if (!m_spectrumNumberTranslator)
+    return 0;
+  return m_spectrumNumberTranslator->globalSize();
 }
 
-/// Return a vector of the detector IDs for given index.
-const std::vector<DetectorID> &
-IndexInfo::detectorIDs(const size_t index) const {
-  return (*m_detectorIDs)[index];
+/// Returns the spectrum number for given index.
+SpectrumNumber IndexInfo::spectrumNumber(const size_t index) const {
+  return m_spectrumNumberTranslator->spectrumNumber(index);
 }
 
 /// Set a spectrum number for each index.
 void IndexInfo::setSpectrumNumbers(
     std::vector<SpectrumNumber> &&spectrumNumbers) {
-  if (size() != spectrumNumbers.size())
+  if (m_spectrumNumberTranslator->globalSize() != spectrumNumbers.size())
     throw std::runtime_error(
         "IndexInfo: Size mismatch when setting new spectrum numbers");
-  m_spectrumNumbers.access() = std::move(spectrumNumbers);
-  m_spectrumNumberTranslator =
-      Kernel::cow_ptr<SpectrumNumberTranslator>{nullptr};
+  makeSpectrumNumberTranslator(std::move(spectrumNumbers));
 }
 
 /// Set a contiguous range of spectrum numbers.
 void IndexInfo::setSpectrumNumbers(const SpectrumNumber min,
                                    const SpectrumNumber max) {
-  if (static_cast<int64_t>(size()) !=
-      static_cast<int32_t>(max) - static_cast<int32_t>(min) + 1)
+  auto newSize = static_cast<int32_t>(max) - static_cast<int32_t>(min) + 1;
+  if (static_cast<int64_t>(m_spectrumNumberTranslator->globalSize()) != newSize)
     throw std::runtime_error(
         "IndexInfo: Size mismatch when setting new spectrum numbers");
```

Comment:
Similarly here. I feel it would be good to indicate a mismatch to the global size.